### PR TITLE
Testing time improvements: Alis/3049

### LINKED ISSUE #3049: Run FE and BE test suites concurrently on PRs
### Description
We'd like to speed up the merge to main process, as there are several automated checks that together take around 10 minutes. One potential solution to speed things up is to run the frontend and backend test suites concurrently, and to explore running the tests themselves concurrently. We think the frontend tests are already run concurrently.

### Proposed solution 
These tests currently run within the same action (`test`) but there is no reason they need to be run together, so we can split them into different actions.

### PR DESCRIPTION
**Pull request by alismx**
# DEVOPS PULL REQUEST

## Related Issue

- Resolves #3049

- This is work to reduce the time it takes to run tests. 
- I took a quick look at testing times, and before this change, I saw time ranges between 17m 4s and 23m 4s. The test runs on this PR are as low as 11m 57s.
- Additionally, now you can rerun just the failed test suite, which cuts down on wasted time from test flake, and sonar will still be able to build a full report due to the cache we make.

## Changes Proposed

- This breaks out backend, frontend, and function tests into different jobs and caches the output required for Sonar to report on the results.

## Additional Information

- I initially tried using artifacts (because that's what I'm more comfortable with), but the gains were minimal. It took too much time to upload and download the artifacts between jobs.

## Testing

- Double-check that the testing and coverage output from the jobs is sufficient for sonar to do proper sonar reporting.
 
### Cache
- [ ] Is the caching sufficient
- [ ] Does the cache names make sense

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -36,15 +36,13 @@ jobs:
         ports:
           - 5432:5432
     steps:
-      - uses: actions/checkout@v2
```
Comment (zdeveloper) [inline .github/workflows/test.yml:39]: I still see a lot of `actions/cache@v2` in the code, can we make sure we either change all of them or none of them, just to keep it consistent

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -60,25 +58,35 @@ jobs:
...
+        with:
+          path: |
+            backend/build/**
+          key: backend-coverage-${{ hashFiles('backend/build/**') }}
+      - name: Archive failed test results
         uses: actions/upload-artifact@v2
         if: failure()
         with:
-          name: test-report
+          name: backtest-report
```
Comment (zdeveloper) [inline .github/workflows/test.yml:77]: can we change this to `backend-test-report`, `backtest` is not very descriptive

**Review by zdeveloper — CHANGES_REQUESTED** [review]
LGTM! with minor nitpicks

**Review by rin-skylight — APPROVED** [review]
This looks good to me! Thank you for the hard work on these improvements.

**Review by zdeveloper — APPROVED** [review]
LGTM! thanks for the updates!
All of the jobs look way faster! Except for the `coverity-scan` takes **18 minutes**! I dont see this job on other PRs, is this a new addition?
image

**Comment by alismx** [comment]
> LGTM! thanks for the updates! All of the jobs look way faster! Except for the `coverity-scan` takes **18 minutes**! I dont see this job on other PRs, is this a new addition? image

@zdeveloper The `coverity-scan` job usually runs on a cron job but it ran against this PR because I updated the file.

```
  push:
    branches:
      - "**"
    paths:
      - .github/workflows/coverity_scan.yml
 ```

**Pull request merged**